presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayHashCodeOperator.java

```diff
@@ -13,48 +13,142 @@
  * limitations under the License.
  */
 
-import com.facebook.presto.metadata.FunctionRegistry;
-import com.facebook.presto.metadata.SqlOperator;
+import com.facebook.presto.spi.PrestoException;
+import com.facebook.presto.spi.StandardErrorCode;
 import com.facebook.presto.spi.block.Block;
 import com.facebook.presto.spi.type.StandardTypes;
 import com.facebook.presto.spi.type.Type;
-import com.facebook.presto.spi.type.TypeManager;
-import com.google.common.collect.ImmutableList;
+import com.facebook.presto.type.SqlType;
+import com.google.common.base.Throwables;
+import io.airlift.slice.Slice;
 
 import java.lang.invoke.MethodHandle;
-import java.util.Map;
 
 import static com.facebook.presto.metadata.OperatorType.HASH_CODE;
-import static com.facebook.presto.metadata.Signature.comparableTypeParameter;
+import static com.facebook.presto.spi.type.TypeUtils.readNativeValue;
 import static com.facebook.presto.type.ArrayType.ARRAY_NULL_ELEMENT_MSG;
 import static com.facebook.presto.type.TypeUtils.checkElementNotNull;
-import static com.facebook.presto.util.Reflection.methodHandle;
 
-public class ArrayHashCodeOperator
-        extends SqlOperator
+@ScalarOperator(HASH_CODE)
+public final class ArrayHashCodeOperator
 {
-    public static final ArrayHashCodeOperator ARRAY_HASH_CODE = new ArrayHashCodeOperator();
-    public static final MethodHandle METHOD_HANDLE = methodHandle(ArrayHashCodeOperator.class, "hash", Type.class, Block.class);
+    private ArrayHashCodeOperator() {}
 
-    private ArrayHashCodeOperator()
+    @TypeParameter("T")
+    @SqlType(StandardTypes.BIGINT)
+    public static long hash(
+            @OperatorDependency(operator = HASH_CODE, returnType = StandardTypes.BIGINT, argumentTypes = {"T"}) MethodHandle hashFunction,
+            @TypeParameter("T") Type type,
+            @SqlType("array(T)") Block block)
     {
-        super(HASH_CODE, ImmutableList.of(comparableTypeParameter("T")), StandardTypes.BIGINT, ImmutableList.of("array(T)"));
+        int hash = 0;
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            checkElementNotNull(block.isNull(i), ARRAY_NULL_ELEMENT_MSG);
+            try {
+                hash = (int) CombineHashFunction.getHash(hash, (long) hashFunction.invoke(readNativeValue(type, block, i)));
```

Comment:
@nezihyigitbasi refactored this in master to use `TypeUtils.hashPosition`. Can you check to see if that has the same performance as having all the different specializations? It might be that you don't need them, if we continue using `TypeUtils.hashPosition`.